e2e/malicious-join/malicious-join.go

```diff
@@ -0,0 +1,207 @@
+/*
+Copyright (c) Edgeless Systems GmbH
+
+SPDX-License-Identifier: AGPL-3.0-only
+*/
+
+// End-to-end test that issues various types of malicious join requests to a cluster.
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/edgelesssys/constellation/v2/internal/attestation/variant"
+	"github.com/edgelesssys/constellation/v2/internal/cloud/cloudprovider"
+	"github.com/edgelesssys/constellation/v2/internal/grpc/dialer"
+	"github.com/edgelesssys/constellation/v2/internal/logger"
+	"github.com/edgelesssys/constellation/v2/joinservice/joinproto"
+	"go.uber.org/zap/zapcore"
+)
+
+// Flags are defined globally as `go test` implicitly calls flag.Parse() before executing a testcase.
+// Thus defining and parsing flags inside a testcase would result in a panic.
+// See https://groups.google.com/g/golang-nuts/c/P6EdEdgvDuc/m/5-Dg6bPxmvQJ.
+var (
+	jsEndpoint = flag.String("js-endpoint", "", "Join service endpoint to use.")
+	csp        = flag.String("csp", "", "Cloud service provider to use.")
+	attVariant = flag.String(
+		"variant",
+		"",
+		fmt.Sprintf("Attestation variant to use. Set to \"default\" to use the default attestation variant for the CSP, or one of: %s", variant.GetAvailableAttestationVariants()),
+	)
+)
+
+func main() {
+	flag.Parse()
+	fmt.Println(formatFlags())
+
+	testCases := map[string]struct {
+		fn      func() error
+		wantErr bool
+	}{
+		"JoinFromUnattestedNode": {
+			fn:      JoinFromUnattestedNode,
+			wantErr: true,
+		},
+	}
+
+	allPassed := true
+	testOutput := &testOutput{
+		TestCases: make(map[string]testCaseOutput),
+	}
+	for name, tc := range testCases {
+		fmt.Printf("Running testcase %s\n", name)
+		err := tc.fn()
+		if tc.wantErr && err == nil {
+			fmt.Printf("Test case %s failed: Expected error but got none\n", name)
+			testOutput.TestCases[name] = testCaseOutput{
+				Passed:  false,
+				Message: fmt.Sprintf("Expected error but got none: %s", err),
```

Comment:
```suggestion
				Message: fmt.Sprintf("Expected error but got none"),
```
There is no error to print here